src/libcore/ffi.rs

```diff
@@ -53,80 +52,158 @@ impl fmt::Debug for c_void {
            reason = "the `c_variadic` feature has not been properly tested on \
                      all supported platforms",
            issue = "44930")]
-extern {
-    type VaListImpl;
-}
+mod impls {
+    use crate::fmt;
+    use crate::marker::PhantomData;
+    use crate::mem;
+    use crate::ops::{Deref, DerefMut};
+    use super::c_void;
 
-#[cfg(any(all(not(target_arch = "aarch64"), not(target_arch = "powerpc"),
-              not(target_arch = "x86_64")),
-          all(target_arch = "aarch64", target_os = "ios"),
-          windows))]
-impl fmt::Debug for VaListImpl {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "va_list* {:p}", self)
+    /// Basic implementation of a `va_list`.
+    #[lang = "va_list"]
+    #[repr(transparent)]
+    pub struct VaListImpl<'f> {
+        ptr: *mut c_void,
+        _marker: PhantomData<&'f c_void>,
     }
-}
 
-/// AArch64 ABI implementation of a `va_list`. See the
-/// [AArch64 Procedure Call Standard] for more details.
-///
-/// [AArch64 Procedure Call Standard]:
-/// http://infocenter.arm.com/help/topic/com.arm.doc.ihi0055b/IHI0055B_aapcs64.pdf
-#[cfg(all(target_arch = "aarch64", not(target_os = "ios"), not(windows)))]
-#[repr(C)]
-#[derive(Debug)]
-#[unstable(feature = "c_variadic",
-           reason = "the `c_variadic` feature has not been properly tested on \
-                     all supported platforms",
-           issue = "44930")]
-struct VaListImpl {
-    stack: *mut c_void,
-    gr_top: *mut c_void,
-    vr_top: *mut c_void,
-    gr_offs: i32,
-    vr_offs: i32,
-}
+    impl<'f> fmt::Debug for VaListImpl<'f> {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "va_list* {:p}", self)
+        }
+    }
 
-/// PowerPC ABI implementation of a `va_list`.
-#[cfg(all(target_arch = "powerpc", not(windows)))]
-#[repr(C)]
-#[derive(Debug)]
-#[unstable(feature = "c_variadic",
-           reason = "the `c_variadic` feature has not been properly tested on \
-                     all supported platforms",
-           issue = "44930")]
-struct VaListImpl {
-    gpr: u8,
-    fpr: u8,
-    reserved: u16,
-    overflow_arg_area: *mut c_void,
-    reg_save_area: *mut c_void,
+    impl<'f> VaListImpl<'f> {
+        /// Convert a `VaListImpl` into a `VaList` that is binary-compatible with C's `va_list`.
+        #[inline]
+        pub fn as_va_list<'s>(&'s self) -> VaList<'s, 'f> {
+            VaList {
+                ptr: self.ptr,
+                _marker1: PhantomData,
+                _marker2: PhantomData,
+            }
+        }
+    }
+
+    /// A wrapper for a `va_list`
+    #[repr(transparent)]
+    #[derive(Debug)]
+    pub struct VaList<'a, 'f: 'a> {
+        ptr: *mut c_void,
```

Comment:
Okay, the transmutes confused me, but basically you can just have a `VaListImpl<'f>` as a field and the `#[repr(transparent)]` should still work.

Reply:
Right, I guess it is cleaner that way, and I can get rid of the transmutes. I'll change it.